Add sub-RFC for increased availability of NUMA API (#1545)
uxlfoundation/oneTBB: rfcs/proposed/numa_support/tbbbind-link-static-hwloc.org

# -*- fill-column: 80; -*-

#+title: Link ~tbbbind~ with static HWLOC to improve predictability of NUMA support API

*Note:* This is a sub-RFC of the https://github.com/oneapi-src/oneTBB/pull/1535.
Specifically, its section about "Increased availability of NUMA support".

* Introduction
oneTBB has a soft dependency on several variants of ~tbbbind~, which are loaded
by the library as part of its initialization stage. In turn, each ~tbbbind~ has
a hard dependency on a concrete version of the HWLOC library [1, 2]. The soft
dependency of oneTBB on ~tbbbind~ allows the library to continue its execution
even if the system loader is unable to resolve the hard dependency on HWLOC for
~tbbbind~. In this case, the HW topology is not discovered and the machine is

> seen as if all CPU cores were uniform, which is the default TBB behavior when

Author: No.

NUMA constraints are not used. Thus, the following code returns the values that
do not reflect the real topology and do not matter:

#+begin_src C++
std::vector<oneapi::tbb::numa_node_id> numa_nodes = oneapi::tbb::info::numa_nodes();
std::vector<oneapi::tbb::core_type_id> core_types = oneapi::tbb::info::core_types();
#+end_src

This lack of valid HW topology data due to absence of a third party library is
the major problem with the current oneTBB behavior. There is no diagnostics for
the issue, which likely makes it unnoticeable by developers, and the code that
uses oneTBB NUMA support facilities continues running but does not use NUMA as
intended.

Having a dependency on a shared HWLOC library has advantages:
1. Code reuse with all of the positive consequences out of this, including

>    relying on the same code that has been tested and debugged, allowing the OS

Reviewer: but in fact most of Linux OSes has obsolete hwloc versions and relying on it does not provide benefits. 
IMO, having most up-to-date static HWLOC together with recent versions of oneTBB has benefits and fixes / new features are available to oneTBB immediately

Author: Thanks, I did not know that. I will consider this in the future changes.

* Proposal
1. Replace the dynamic link of ~tbbbind~ library which is currently linked
   against HWLOC 1.x with the link to a static HWLOC library version 2.x.
2. Add loading of that ~tbbbind~ variant as the last attempt to resolve the
   dependency on functionality provided by ~tbbbind~ layer.
3. Update the oneTBB documentation considering [[https://oneapi-src.github.io/oneTBB/search.html?q=tbb%3A%3Ainfo][these documentation pages]] to
   include steps determining the variant of ~tbbbind~ being used.

** Advantages
1. The proposed behavior allows having a mechanism for resolving a dependency on
   HWLOC library in case it cannot be found in the environment, while still
   preferring user-provided version of HWLOC. As a result, the problematic use of
   oneTBB API mentioned above should work as expected, returning enumerated list
   of actual NUMA nodes and core types on the system the code is running on,
   provided that the loaded HWLOC library works on that system and that an
   application properly distributes all binaries of oneTBB, sets the environment
   so that the necessary variant of ~tbbbind~ library can be found and loaded.

> 2. The drop of support for HWLOC 1.x allows to not introducing additional

Author: ```suggestion
2. Dropping support for HWLOC 1.x, does not introduce an additional
```

   ~tbbbind~ variant of the library, yet maintaining support for popular

>    versions of HWLOC.

Reviewer: actually, you still introduce new variant of tbbbind, as ALL distributions already ship only libtbbbind_2_5.so

It means, all TBB distribution across all package managers need to be updated

Author: Or they may continue supporting the current variant with system-specific version of HWLOC as it is still a working approach.

** Disadvantages
By default still no diagnostics if users failed to setup environment with their

> own version of HWLOC library correctly. Although, specifying ~TBB_VERSION=1~

Reviewer: why user would be its own version of hwloc?

Author: We introduced that approach in the past and it may be backward incompatible if we change it at once. Perhaps, we decide to prioritize own version of HWLOC someday, but AFAIK still need to leave the possibility to specify a user one.

envar will help identifying an issue with setup of environment pretty quickly.

* Alternative handling of inability to parse system topology

> The other behavior in case HWLOC library cannot be found is to be more explicit

Reviewer 2: An alternative approach to handle the absence of the HWLOC library is to adopt a more explicit response:
* Issue a warning about the missing component.
* Require one of the tbbbind variants to be loaded by refusing to work or throwing an exception.

about the problem of a missing component and to either issue a warning or to
refuse working requiring one of the ~tbbbind~ variant to be loaded (e.g., throw
an exception).

> Comparing these alternative approaches to the one proposed.

Reviewer 2: should this be a heading?

Author: No necessarily.

** Common Advantages

> - Explicitly tells that the functionality being used is not going to work

Reviewer 2: ```suggestion
- Explicitly indicates that the functionality being used does not work,
```

>   instead of just being silent.

Reviewer 2: ```suggestion
  instead of failing silently.
```

> - Does not require additional variant of ~tbbbind~ library to be distributed

Reviewer 2: ```suggestion
- Avoids the need to distribute an additional variant of ~tbbbind~ library. 
```

>   along with the others.

Reviewer 2: ```suggestion

```

** Common Disadvantages
- Requires additional step from the user side to resolve the problem. In other
  words, it does not provide complete solution to the problem.

> ** Disadvantages of Issuing a Warning

Reviewer 2: should this section and the next one be subparts of the common disadvantages section?

Author: Yes, let's make it that way.
```suggestion
*** Disadvantages of Issuing a Warning
```

> - The warning may still not be visible, especially if standard streams are

Reviewer 2: ```suggestion
- The warning may be unnoticed, especially if standard streams are
```

  closed.

> ** Disadvantages of Throwing an Exception

Author: ```suggestion
*** Disadvantages of Throwing an Exception
```

> - May break existing code as it does not expect an exception to be thrown.

Reviewer 2: ```suggestion
- May break existing code that does not expect an exception to be thrown.
```
